### ISSUE TITLE
Cascading Card Art Granularity

### ISSUE THREAD

**Issue by emgreen8863** [Cockatrice/Cockatrice#2790]
<b>OS:</b>
Win 10

<b>Cockatrice version:</b>
2.3.17 (2017-05-05)
__________

Currently the art used for a given card is determined via the following priority:

```
1. The CUSTOM Folder
2. Enabled Sets (Top to Bottom)
3. Disabled Sets (Top to Bottom)
```

**The Problem**

This limits the options to specify the art used for a single card to the use of a custom set, and the process for doing so is more complicated than it perhaps could be otherwise.  In addition to this the change is made globally rather than on a per deck basis, which at least to me, would be more valuable.

It wouldn't make sense to change the format of the deck save files to achieve this as it would create legacy compatibility issues and other problems when exporting or importing to other applications.  Instead, why not apply the principles of CSS?

--------------------------------------------

**Changes to the Model**

Each deck may optionally have a style sheet that may dictate set priority or specific art on a per card basis.  This carries higher priority than global set listings.  As a result the art used would be determined in the following order:

```
1. Per card art definitions at the deck level
2. Set priority determined at the deck level
3. The CUSTOM Folder
4. Enabled Sets (Top to Bottom)
5. Disabled Sets (Top to Bottom)
```

The new deck style sheet would likely have two entry types similar in format to those found below:

```
set <setName>
```

```
cardName <setName>
```

Lines containing ```set ``` definitions would create a set priority in the order they appear.  **This does not need to be a full set listing.**  By that I mean that you should be free to specify only one or two sets and to allow the priority of any remaining cards to be determined by the global priority.  Lines containing definitions for individual cards should apply though preferences accordingly.

--------------------------------------------

**Changes to the View**

This framework should allow changes to the view or GUI that make the process of selecting specific art much more intuitive.  In this area I make three suggestions.

The first of these is to include left and right arrow buttons on either side of the card preview in the deck editor.  These should allow scrolling through the available sets/art for the selected card and will save the changes to the deck style sheet when the deck is saved.  This should greatly streamline this process for end users.

The second suggestion is to include a set priority button in the deck editor to add and define set priorities at the deck level.  This too will save the changes to the deck style sheet when the deck is saved.

The third suggestion addresses issue #2022 at least in part.  A feature currently exits that allows the creation of a related card.  I would suggest the reuse of this code in the deck editor.  Assuming it executes pretty quickly you could run it after each change to the deck list *(card name added or removed from list, non-duplicate)* to keep a running list of potential tokens used by the deck.  This list should not be editable, and should exist as a separate entry in the tree like the sideboard.  This will allow the use of the arrow buttons from the first suggestion to apply specific art for a given token.  I feel like this is the best implementation for this feature, and understand that it may require additional work regarding how tokens are currently stored.

--------------------------------------------

**Other Thoughts**

This approach does have some issues when trying to implement the suggestion from #486 but I have some ideas as to how to mitigate those problems.  In addition to this I feel it might be nice to allow multiple arts for both lands and tokens, or for that matter any card.

Implementing this with this solution would need to allow for an number parameter in the style sheet.  For example:

```
3 island HOU
```

In this case we're indicating that 3 islands should be given the art from the Hour of Devastation Set.  In reality it's a bit more complicated than this because there may be more than one art per set, so instead an identifier for the art will need to be used.

It's possible that conflicts between the deck style file and the desk list occur as a result of the duplicate numeric lists that must be updated in the model, but they can be addressed easily enough.  Should a case arise in which there are more entries for a card or card in the style sheet than there are in the deck list any art definitions that exceed the number listed in the deck list should be ignored.

Finally, how does this all work out at play time when players are actually drawing cards?  I haven't looked at the current architecture but I don't believe it will be necessary to change the current deck model to account for this.  A two dimensional array (or array of arrays) that accompanies the deck could solve this problem.  Each array in the array will correspond with a card for which more than one art is present and will contain entries that correspond with artID's.  Cards can be drawn as normal.  Should you encounter a card with multiple potential arts an entry with an artID will be removed from the appropriate array at random and the corresponding art will be used.

--------------------------------------------

**A Crazy Notion**

It would be awesome if we could share deck style sheets so that other users see our decks the same way that we do.  I get that there are problems with doing so.

The style sheets would likely need to be encrypted and perhaps require a key from the server so that players couldn't gain information about opposing decks.  In addition to this problems such as the one described in #2330 will need to be accounted for.  That said so long as translations appear correctly elsewhere it might be OK to allow art with other translations.  Beyond that to address this problem a setting could be added to disable this feature entirely or style sheets including art for another region.

In short this particular suggestion might very well be impracticable or not feasible, but it's a cool thought all the same.

--------------------------------------------

**In Closing**

While implementing a framework of this nature will likely require some rework of the existing source, the hope is that this can mostly be tacked on top of the current base.  I feel that this would greatly improve the current feature set regarding the use of specific art, and improve the presence of deck flavor and styling, and improve the user experience.  I haven't done a ton of work in C++ specifically, but feel free to hit me up should you want some help implementing such a feature set.  If nothing else I could use the padding for my resume.

**Comment by skwerlman** [comment]
Linking #841 #1847 #2545

This solution has some issues. For example, take the line `1 RULE OF LAW`. If Wizards were ever to release a set with the code `LAW`, Cockatrice would be unable to tell if the user wanted the card "Rule of Law" or if they wanted the card "Rule of", preferring art from the set `LAW`. This is because the copy-paste format is case-insensitive, and allows unknown cards to be included.

IMO, it'd be better to extend [the .cod format](https://github.com/Cockatrice/Cockatrice/blob/master/doc/deck.xsd) to allow specifying card art instead, since it already supports other features (like deck descriptions) which aren't present in other deck formats.

**Comment by emgreen8863** [comment]
Interesting...

To correct the problem you've mentioned you could simply use a different delimiter such as a colon or a comma.  I actually had colons in there originally and removed them.  I suppose extending the .cod format wouldn't cause any compatibility issues as the style tags or whatever you'd like to call them would be optional in nature.  It would remove the duplicate update as well.

I thought I did a decent job of searching for existing issues previously so apologies for creating a new ticket if it was not warranted.  Perhaps my proposition was ill considered.

**Comment by skwerlman** [comment]
Nah, new ideas are good, esp. since this is a problem which hasn't been thought about in a while.

I like the idea of using a delimiter to separate out the set code, but we should probably use a char sequence which cannot (within reason) appear on a card, like `;;` or something similar.

**Comment by emgreen8863** [comment]
Ah yeah that's true, it's pretty conceivable that Wizards could include a lone colon in a card name.  In fact they have with all of the Circle of Protection's.  Let me know if you find any other issues with it.  I feel like determining the art used on a draw in game could be done more elegantly, but I still haven't gotten around to looking at the source.  Are all cards objects?  If so I guess it wouldn't be too hard to add another attribute called artID or something similar.  It would definitely make more sense than my original convoluted suggestion.

**Comment by Thunderbeliever** [comment]
Personally i use Custom all the time and having it be the Last option would be a pain

**Comment by emgreen8863** [comment]
Thunderbeliever, can you elaborate?  Custom sets would still have the highest global priority.  I fail to see how this would impact their use.

**Comment by Thunderbeliever** [comment]
It sounds like you want to make Custom the Last option and that means that it would have to be reset every time. If I am misreading this then i apologize

**Comment by skwerlman** [comment]
no, this proposal allows the optional specification of a set code in a deck list so cockatrice can display multiple card arts per card. Custom sets are treated the same way as normal sets by it.

**Comment by ZeldaZach** [comment]
This is heavily related (and dupe) to #217
